iceoryx_hoofs/posix/ipc/include/iox/detail/unix_domain_socket.inl

```diff
@@ -0,0 +1,160 @@
+// Copyright 2023, Eclipse Foundation and the iceoryx contributors. All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+// SPDX-License-Identifier: Apache-2.0
+#ifndef IOX_HOOFS_POSIX_IPC_UNIX_DOMAIN_SOCKET_INL
+#define IOX_HOOFS_POSIX_IPC_UNIX_DOMAIN_SOCKET_INL
+
+#include "iox/posix_call.hpp"
+#include "iox/unix_domain_socket.hpp"
+
+namespace iox
+{
+template <typename Type>
+expected<void, PosixIpcChannelError> UnixDomainSocket::sendImpl(const Type* msg,
+                                                                uint64_t msgSize,
+                                                                const units::Duration& timeout,
+                                                                const Termination terminator) const noexcept
+{
+    if (msgSize > m_maxMessageSize)
+    {
+        return err(PosixIpcChannelError::MESSAGE_TOO_LONG);
+    }
+
+    if (PosixIpcChannelSide::SERVER == m_channelSide)
+    {
+        IOX_LOG(ERROR, "sending on server side not supported for unix domain socket \"" << m_name << "\"");
+        return err(PosixIpcChannelError::INTERNAL_LOGIC_ERROR);
+    }
+
+    auto tv = timeout.timeval();
+    auto setsockoptCall = IOX_POSIX_CALL(iox_setsockopt)(m_sockfd, SOL_SOCKET, SO_SNDTIMEO, &tv, sizeof(tv))
+                              .failureReturnValue(ERROR_CODE)
+                              .ignoreErrnos(EWOULDBLOCK)
+                              .evaluate();
+
+    if (setsockoptCall.has_error())
+    {
+        return err(errnoToEnum(setsockoptCall.error().errnum));
+    }
+    uint64_t msgSizeToSend = msgSize;
+    if (terminator == Termination::NULL_TERMINATOR)
```

Comment:
The ```terminator``` value is not a constexpr. 
I also tried with the compiler to be sure, and this is the error I got:
```
In file included from /home/lbartoli/repos/iceoryx/iceoryx_hoofs/posix/ipc/include/iox/unix_domain_socket.hpp:221,
                 from /home/lbartoli/repos/iceoryx/iceoryx_hoofs/posix/ipc/source/unix_domain_socket.cpp:19:
/home/lbartoli/repos/iceoryx/iceoryx_hoofs/posix/ipc/include/iox/detail/unix_domain_socket.inl: In member function ‘iox::expected<void, iox::PosixIpcChannelError> iox::UnixDomainSocket::timedSendImpl(const Type*, uint64_t, const iox::units::Duration&, iox::UnixDomainSocket::Termination) const’:
/home/lbartoli/repos/iceoryx/iceoryx_hoofs/posix/ipc/include/iox/detail/unix_domain_socket.inl:53:61: error: ‘terminator’ is not a constant expression
   53 |     if constexpr (terminator == Termination::NULL_TERMINATOR)
      |
```

Reply:
Oh, right. `Termination` must become a template parameter.

Reply:
My suggestion is to keep it as is, the function signature is simple, otherwise, when you call the method you need to pass the type and the terminator as a template parameter, overcomplicating the function call.

I also see no better performance to move in a template parameter: the function is a template and is not compiled and linked so that the terminator will be hardcoded to ```false``` or ```true```, the compiler will do the same optimization.

What do you think about it?

Reply:
Oh, I was not clear enough. I meant a non-type template parameter. The usage would then be `timedSendImpl<Termination::NULL_TERMINATOR>(msg, size, timeout)`. But you are right, the compiler should be able to optimize this anyway. There might be some tools which don't get it right and complain but that is something for our future selves. I'll leave it to you.

Reply:
Specified the template parameter in the function call the code works

Reply:
Given that I need to specify the msg type to use the```not_null``` constructor, I also added it to the template parameter list